upstairs/src/upstairs.rs

```diff
@@ -437,8 +474,13 @@ impl Upstairs {
             UpstairsAction::Control(c) => {
                 self.on_control_req(c).await;
             }
+            UpstairsAction::NoOp => (),
         }
 
+        // Check whether we need to mark a Downstairs as faulted because too
+        // many jobs have piled up.
+        self.gone_too_long();
```

Comment:
Yeah, the `Guest` action will now either
- Immediately handle the `BlockReq` (if it's not a write _and_ there are no deferred actions in the queue), **or**
- Defer that `BlockReq` into the queue (if it's a write, _or_ other `BlockReq`'s are already deferred)

We'd want to check `gone_too_long` in the first case but not the second; in the second case, we'd want to check `gone_too_long` when the deferred `BlockReq` is resolved (from `UpstairsAction::DeferredBlockReq`).

Instead of trying to keep track of this, I opted to move `gone_too_long` into the "invariant checking" section of `Upstairs::apply`.  It should be a cheap O(1) check, so this seemed like the simplest option.